docs/references/architecture/adr-113-modular-transaction-hashing.md

```diff
@@ -0,0 +1,154 @@
+# ADR 113: Modular transaction hashing
+
+## Changelog
+
+- 2024-02-05: First version (@melekes)
+- 2024-05-28: Complete refactor (@melekes)
+
+## Status
+
+Proposed
+
+## Context
+
+Transaction hashing in CometBFT is currently implemented using `crypto/tmhash`
+package, which itself relies on [`sha256`](https://pkg.go.dev/crypto/sha256) to
+calculate a transaction's hash.
+
+The hash is then used by:
+
+- the built-in transaction indexer;
+- the `/tx` and `/tx_search` RPC endpoints, which allow users
+to search for a transaction using its hash;
+- `types` package to calculate the Merkle root of block's transactions.
+
+`tmhash` library is also used to calculate various hashes in CometBFT repo (e.g.,
+evidence, consensus params, commit, header, partset header).
+
+The problem some application developers are facing is a mismatch between the
+internal/app representation of transactions and the one employed by CometBFT. For
+example, [Evmos](https://evmos.org/) wants transactions to be hashed using
+the [RLP][rlp].
+
+In order to be flexible, CometBFT needs to allow changing the transaction
+hashing algorithm if desired by the app developers.
+
+## Alternative Approaches
+
+1. Do nothing => not flexible.
+2. Add `HashFn` argument to `NewNode` and pass this function down the stack =>
+   complicates the code.
+3. Limit the scope of the solution described below to just transaction hashing
+   (do not change header's hash, evidence's hash, etc.) => ?
+
+## Decision
+
+Give app developers a way to provide their own transaction hash function.
+
+## Detailed Design
+
+Use `sha256` by default, but give developers a way to change the hashing function:
+
+```go
+import (
+	"crypto"
+    "hash"
+	"crypto/sha256"
+)
+
+const (
+	// The truncated size of a checksum in bytes.
+	TruncatedSize = 20 // validators addresses
+)
+
+var (
+
+    // Hash used
+    Hash = crypto.SHA256
+
+	stringFunc = func(bz []byte) string {
+		return fmt.Sprintf("%X", bz)
+	}
```

Comment:
Yes, I did. I found that having two separate things (object and a function is a cleaner approach)

Reply:
Fair enough

Reply:
I agree, but the problem is that the standard `hash.Hash` https://pkg.go.dev/hash#Hash doesn't have a `String` method, so this will require our own interface, meaning we won't be able to use `crypto.Hash` from the Go stdlib. Again, nothing is wrong, but ideally, we'd use standard interfaces.